### DISCUSSION TITLE
How to set background color for insert_htmlbox(rect, text, *, css=None, scale_low=0, archive=None, rotate=0, oc=0, opacity=1, overlay=True)?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#4102] [Q&A]
How to set background color for insert_htmlbox(rect, text, *, css=None, scale_low=0, archive=None, rotate=0, oc=0, opacity=1, overlay=True)?

div code and span code did not work for style="background-color: aqua;"

**Comment by nissansz** [comment]
This does not work: background-color:rgba(0, 255, 255, 0.7);  
This works: background-color:yellow;

**Reply by JorjMcKie** [reply]
Your text must contain HTML/CSS setting background color with an HTML tag supporting this, like `<p>` or  `<div>`.
However, coloring only happens for the sub-rectangle actually used by the text.

**Comment by nissansz** [comment]
![image](https://github.com/user-attachments/assets/a0c6511a-2600-4373-b945-3541389560df)

Any html code or pymupdf code to stretch the text in html box to be fully fill the box evenly instead of no spacing?

Want to match previous textwriter effect.

![image](https://github.com/user-attachments/assets/04012d7f-a03c-4554-9035-e740e3dfa8ac)

**Reply by JorjMcKie** [reply]
Not sure. Not the full HTML / CSS syntax is supported by MuPDF. Only HTML4 / CSS2 - and with exceptions.
Try this: https://developer.mozilla.org/en-US/docs/Web/CSS/font-stretch

**Reply by JorjMcKie** [reply]
No, apparently no support of font-stretch.

**Comment by nissansz** [comment]
How to support below html effect for "text-align: justify;text-align-last:justify;? It seems supported by HTML4 / CSS2
![image](https://github.com/user-attachments/assets/3cf8cd65-77ee-4aeb-9080-748fa487115d)

**Reply by JorjMcKie** [reply]
Don't know. You have to try out if supported by muPDF.

**Reply by nissansz** [reply]
It does not work.

**Comment by nissansz** [comment]
Is it possible to set color for each character in a textline for textwriter?

**Reply by JorjMcKie** [reply]
only with different textwriters, one for each color

**Comment by nissansz** [comment]
I tried to use below textwriter for each character.
But I failed to move the character to the yellow box area, code as below.  
Also, rotate parameter throws error in textwriter.
Is my usage correct to put tw = fitz.TextWriter(page.rect) in the for loop? Or, can it be defined as one textwriter only before for loop?
'''

            half_font = int(fontsize_fit) / 2

            total_count = 0
            g = 0
            updateleft=0
            for char in character_set:
                    total_count += 1  # 英文、数字或标点符号算作 0.5
                    tw = fitz.TextWriter(page.rect)
                    origin = (int(updateleft), int(bottom+font.descender*(bottom-top)))
                    pos = fitz.Point(updateleft, top-half_font)
                    mat = fitz.Matrix(half_font / fontsize, 1.0)
                    tw.append(origin, clean_char, fontsize=half_font, font=font, small_caps=True)

                    if  'aqua' in char:
                        color =  (0 / 255, 0 / 255, 255 / 255)
                    elif  'red' in char:
                        color =  (255 / 255, 0 / 255, 0 / 255)
                    else:
                        color =  (0 / 255, 0 / 255, 0 / 255)
                    tw.write_text(page, opacity=None, color=color, morph=(pos, mat), overlay=True, oc=0, render_mode=0)
                    # page.insert_text(origin, clean_char, fontsize=half_font, fontfile= font ,
                    #              color=color, fill=None, render_mode=0, border_width=1, rotate=rotate, morph=(pos, mat),
                    #              stroke_opacity=1, fill_opacity=1, overlay=True, oc=0)

'''

![image](https://github.com/user-attachments/assets/2a281061-a278-43df-ba33-9d32d3b7de8d)

**Comment by nissansz** [comment]
Some text like 0, 2 in the first line is very small, I used local font Deng.
Still cannot stretch to fit whole rect. 
How to fill / calculate mat for morph?

![image](https://github.com/user-attachments/assets/3af5a1e2-678a-44f7-9684-a7df78b4b81b)
